internal/docstore/mongodocstore/mongo.go

```diff
@@ -0,0 +1,215 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mongodocstore provides an implementation of the docstore API for MongoDB.
+//
+// Docstore types not supported by the Go mongo client, go.mongodb.org/mongo-driver/mongo:
+// TODO(jba): write
+//
+// MongoDB types not supported by Docstore:
+// TODO(jba): write
+package mongodocstore // import "gocloud.dev/internal/docstore/mongodocstore"
+
+// TODO(jba): revision field
+
+// MongoDB reference manual: https://docs.mongodb.com/manual
+// Client documentation: https://godoc.org/go.mongodb.org/mongo-driver/mongo
+//
+// The client methods accept a document of type interface{},
+// which is marshaled by the go.mongodb.org/mongo-driver/bson package.
+
+import (
+	"context"
+	"strings"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/internal/docstore"
+	"gocloud.dev/internal/docstore/driver"
+	"gocloud.dev/internal/gcerr"
+)
+
+type collection struct {
+	coll *mongo.Collection
+}
+
+type Options struct {
+}
+
+// OpenCollection opens a MongoDB collection for use with Docstore.
+func OpenCollection(mcoll *mongo.Collection, _ *Options) *docstore.Collection {
+	return docstore.NewCollection(newCollection(mcoll))
+}
+
+func newCollection(mcoll *mongo.Collection) *collection {
+	return &collection{coll: mcoll}
+}
+
+// From https://docs.mongodb.com/manual/core/document: "The field name _id is
+// reserved for use as a primary key; its value must be unique in the collection, is
+// immutable, and may be of any type other than an array."
+const idField = "_id"
+
+// TODO(jba): use bulk RPCs.
+func (c *collection) RunActions(ctx context.Context, actions []*driver.Action) (int, error) {
+	for i, a := range actions {
+		if err := c.runAction(ctx, a); err != nil {
+			return i, err
+		}
+	}
+	return len(actions), nil
+}
+
+func (c *collection) runAction(ctx context.Context, action *driver.Action) error {
+	var err error
+	switch action.Kind {
+	case driver.Get:
+		err = c.get(ctx, action)
+
+	case driver.Create:
+		err = c.create(ctx, action)
+
+	case driver.Replace, driver.Put:
+		err = c.replace(ctx, action, action.Kind == driver.Put)
+
+	case driver.Delete:
+		err = c.delete(ctx, action)
+
+	case driver.Update:
+		err = c.update(ctx, action)
+
+	default:
+		err = gcerr.Newf(gcerr.Internal, nil, "bad action %+v", action)
+	}
+	return err
+}
+
+func (c *collection) get(ctx context.Context, a *driver.Action) error {
+	// TODO(jba): use Projection option to return only desired field paths.
+	id, err := a.Doc.GetField(idField)
+	if err != nil {
+		return err
+	}
+	res := c.coll.FindOne(ctx, bson.D{{"_id", id}})
+	if res.Err() != nil {
+		return res.Err()
+	}
+	var m map[string]interface{}
+	// TODO(jba): Consider if we can avoid the double decode here.
```

Comment:
We might want to benchmark the double encoding/decoding first. It seems we need to call its `Decode` anyways?

Reply:
Good point. I changed the TODO.

There is something called `bson.Raw` which I think would cause the decoder to do nothing, and then we could decode the BSON ourselves. It would definitely be a lot of work though.